design.md

```diff
@@ -0,0 +1,247 @@
+  - /sys/fs/cgroup/teleport/<job_id> -- the job_id is the UUID provided by the server when the job was created.
+
+The hardcoded resource limits are below. For this challenge, I will assume the jobs are relatively light: 
+- CPU: 0.5 
+- Memory: 256MB
+- IO: 1MB/s
+
+#### Placing jobs in cgroups
+
+- os.Open opens /sys/fs/cgroup/teleport/<job_id>, resulting in a file descriptor.
+- When running exec.Cmd, use UseCgroupFD and CgroupFD to specify that the process should be placed in a cgroup.
+
+
+#### Tradesoffs and the Production environment
+- It is preferable to dynamically configure resource limits based on job and system state. A first step in doing 
+  this is providing a way to adjust resource limits when jobs are created.
+- In production, we probably want orchestrators, like systemd, which manages cgroups. 
+
+## API
+
+### Using gRPC for communication. 
+
+- Start: Starts a job with hardcoded resource limits.
+- Stop: Stops a job.
+- GetStatus: Fetches status of a job
+- GetLogs: Stream logs of a running job.
+
+```
+service TeleportService {
+    rpc Start (StartRequest) returns (StartResponse);
+    rpc Stop (StopRequest) returns (StopResponse);
+    rpc GetStatus (GetStatusRequest) returns (GetStatusResponse);
+    rpc GetLogs (GetLogsRequest) returns (stream GetLogsResponse);
+}
+
+message StartRequest {
+    string command = 1;
+    repeated string args = 2;
+}
+
+message StartResponse {
+    string job_id = 1; // a job UUID
+}
+
+message StopRequest {
+    string job_id = 1;
+}
+
+message StopResponse {
+    // ideally, add a boolean to see if stopping was successful
+    // maybe even a message. But not necessary for the challenge.
+}
+
+message GetStatusRequest {
+    string job_id = 1;
+}
+
+message GetStatusResponse {
+    Status status = 1;
+}
+
+enum Status{
+    STATUS_UNSPECIFIED = 0; 
+    STATUS_RUNNING = 1;
+    STATUS_STOPPED = 2; // User stopped the job
+    STATUS_EXITED = 3;  // Job exited
+    STATUS_UNKNOWN = 4; // For anything we don't recognize
+}
+
+message LogsRequest {
+  string job_id = 1;
+}
+
+message LogsResponse {
+  bytes data = 1; // log data
+}
+```
+
+### TLS
+
+API will use TLS 1.3, which is the latest version. 1.3 offers improved security and performance gains the previous 
+versions. 
+
+Recommended curves: 
+- X25519 : Performant and secure.
+- P-256 : Widely supported and TLS 1.2 compatible.
+- P-384 : Higher level of security. Good for sensitive data.
+- P-521 : Very high security.
+
+Recommended Cipher suites. These are performant and offers good security:
+- TLS_AES_128_GCM_SHA256
+- TLS_CHACHA20_POLY1305_SHA256
+- TLS_AES_256_GCM_SHA384
+
+### Authentication
+
+mTLS will be used for authentication. Both server and client provides certificates in order to 
+communicate. For testing, premade certificates will be provided.
+
+Users are authorized to use the service by its certification being validated. Validation is done by looking at the 
+Common Name (CN), which is part of the subject.
```

Comment:
I was thinking of simply having a map between user in the CN to a job id. This would indicate what jobs a user can manage. Would that work for this challenge?

Reply:
added: https://github.com/rexposadas/teleport/pull/1/commits/ac600de1755067c600f5afd45279d39176e86e0b

Reply:
Done